File: GundamGame/gameserver/src/main/player.go
package main

type PlayerState int

// const (
// 	Idle PlayerState = iota
// 	Backward
// 	FrontKick
// 	Jump
// 	LowKick
// 	Walking
// )

type Player struct {
	X                   int
	Y                   int
	Width               int
	Height              int
	XSpeed              int
	YSpeed              int
	Health              int
	Type                string
	MaxHealth           int
	State               string
	Id                  string
	Lock                bool
	Mustlock            bool
	LastFrameChangeTime float64
	CurrentFrame        int
}

type PlayerJSON struct {
	X         int    `json:"x"`
	Y         int    `json:"y"`
	Width     int    `json:"width"`
	Height    int    `json:"height"`
	XSpeed    int    `json:"xSpeed"`
	YSpeed    int    `json:"ySpeed"`
	Health    int    `json:"health"`
	Type      string `json:"type"`
	MaxHealth int    `json:"maxHealth"`
	Id        string `json:"id"`
	State     struct {
		State               string  `json:"state"`
		Lock                bool    `json:"lock"`
		MustLock            bool    `json:"mustLock"`
		CurrentFrame        int     `json:"currentFrame"`
		FrameDuration       int     `json:"frameDuration"`
		LastFrameChangeTime float64 `json:"lastFrameChangeTime"`
	} `json:"state"`
}

func (p *Player) SetX(x int) {
	p.X = x
}

func (p *Player) SetY(y int) {
	p.Y = y
}

func (p *Player) SetWidth(width int) {
	p.Width = width
}

func (p *Player) SetHeight(height int) {
	p.Height = height
}

func (p *Player) SetXSpeed(xSpeed int) {
	p.XSpeed = xSpeed
}

func (p *Player) SetYSpeed(ySpeed int) {
	p.YSpeed = ySpeed
}

func (p *Player) SetHealth(health int) {
	p.Health = health
}

func (p *Player) SetMaxHealth(maxHealth int) {
	p.MaxHealth = maxHealth
}

func (p *Player) SetState(state string) {
	p.State = state
}

func (p *Player) SetLock(lock bool) {
	p.Lock = lock
}

func (p *Player) SetMustLock(mustlock bool) {
	p.Mustlock = mustlock
}

func (p *Player) SetLastFrameChangeTime(lastFrameChangeTime float64) {
	p.LastFrameChangeTime = lastFrameChangeTime
}

func (p *Player) SetCurrentFrame(currentFrame int) {
	p.CurrentFrame = currentFrame
}

File: GundamGame/src/script/player/player.ts
import InputState, { isKeyPressed } from "../input/input";
import { game } from "../main";
import SocketMessage from "../socket/socketMessage";
import PlayerState, { State } from "./playerState";

type Player = {
    x: number,
    y: number,
    type: string,
    width: number,
    height: number,
    xSpeed: number,
    ySpeed: number,
    health: number,
    maxHealth: number,
    id: string,
    mirrored: boolean,
    position: string,
    state: PlayerState
}

export const updatePlayerAction = (player: Player, inputState: InputState): Player => {
    let newX = player.x;
    let newY = player.y;
    console.log(player.type);
    let state: PlayerState = player.state.lock ? {
        ...player.state
    }
    : {
        state: player.state.state,
        animations: player.type === "blast_impulse" ?
            player.mirrored ? game.firstPlayerAnimationMapMirrored[player.state.state] : game.firstPlayerAnimationMap[player.state.state] : 
            player.mirrored ? game.secondPlayerAnimationMap[player.state.state] : game.secondPlayerAnimationMapMirrored[player.state.state],
        frameDuration: 250,
        currentFrame: player.state.currentFrame,
        lastFrameChangeTime: player.state.lastFrameChangeTime,
        mustLock: false,
        lock: false
    };

    let width: number = player.width;
    console.log("hehe")
    

    if (isKeyPressed(inputState, 'a')) {
        game.socket.sendMessage("command: " + JSON.stringify({playerID: player.id, command: "moveleft"} as SocketMessage));
        console.log("Send backward message");
        // if (!player.state.lock) {
        //     newX = player.x < 50 ? player.x : player.x - player.xSpeed;
        //     width = 600;
        //     const newDuration = player.state.state === State.Backward ? player.state.currentFrame : 0;
        //     state = !player.state.lock ? {
        //         state: State.Backward,
        //         animations: firstPlayerAnimationMap[State.Backward],
        //         currentFrame: nefirst_playerwDuration,
        //         frameDuration: 250,
        //         lastFrameChangeTime: player.state.lastFrameChangeTime,
        //         mustLock: player.state.mustLock,
        //         lock: player.state.lock
        //     } : player.state;
        // }
    } else if (isKeyPressed(inputState, 'd')) {
        game.socket.sendMessage("command: " + JSON.stringify({playerID: player.id, command: "moveright"} as SocketMessage));
        console.log("Send walking message")
        // if (!player.state.lock) {
        //     newX = player.x > 1400 ? player.x : player.x + player.xSpeed;
        //     width = 600;
        //     const newDuration = player.state.state === State.Walking ? player.state.currentFrame : 0;
        //     state = !player.state.lock ? {
        //         state: State.Walking,
        //         animations: firstPlayerAnimationMap[State.Walking],
        //         currentFrame: newDuration,
        //         frameDuration: 250,
        //         lastFrameChangeTime: player.state.lastFrameChangeTime,
        //         mustLock: player.state.mustLock,
        //         lock: player.state.lock
        //     } : player.state;
        // }   
    } else if (isKeyPressed(inputState, 'j')) {
        game.socket.sendMessage("command: " + JSON.stringify({playerID: player.id, command: "attack"} as SocketMessage));
        // if (!player.state.lock) {
        //     const newDuration = player.state.state === State.FrontKick ? player.state.currentFrame : 0;
        //     width = 700;
        //     state = !player.state.lock ? {
        //         state: State.FrontKick,
        //         animations: firstPlayerAnimationMap[State.FrontKick],
        //         currentFrame: newDuration,
        //         frameDuration: 350,
        //         lastFrameChangeTime: player.state.lastFrameChangeTime,
        //         mustLock: player.state.mustLock,
        //         lock: true
        //     }: player.state;
        // }
    } else if (game.first_player.state.state  !== State.Idle && !player.state.lock) {
        game.socket.sendMessage("command: " + JSON.stringify({playerID: player.id, command: "idle"} as SocketMessage));
        console.log("Send idle message")
    }

    width = state.state === State.Idle ? 400 : width;
    

    return { ...player, width: width, state: state, x: newX }
}

export default Player;